sycl/source/detail/graph_impl.cpp

```diff
@@ -226,6 +226,141 @@ bool graph_impl::clearQueues() {
   return AnyQueuesCleared;
 }
 
+static bool
+markNodeAsVisited(std::shared_ptr<node_impl> Node,
+                  std::vector<std::shared_ptr<node_impl>> &Visited) {
+  assert(Node && "Node can't be nullptr");
+  if (Node->MVisited)
+    return false;
+  Node->MVisited = true;
+  Visited.push_back(Node);
+  return true;
+}
+
+static void
+unmarkVisitedNodes(std::vector<std::shared_ptr<node_impl>> &Visited) {
+  for (std::shared_ptr<node_impl> Node : Visited)
+    Node->MVisited = false;
+}
+
+static void printDotRecursive(std::fstream &Stream,
+                              std::vector<std::shared_ptr<node_impl>> &Visited,
+                              std::shared_ptr<node_impl> Node) {
+  if (!markNodeAsVisited(Node, Visited))
+    return;
+
+  Node->MCommandGroup->printDot(Stream);
+  for (const auto &Dep : Node->MPredecessors) {
+    auto NodeDep = Dep.lock();
+    Stream << "  \"" << Node->MCommandGroup.get() << "\" -> \""
+           << NodeDep->MCommandGroup.get() << "\"" << std::endl;
+  }
+
+  for (std::shared_ptr<node_impl> Succ : Node->MSuccessors) {
+    printDotRecursive(Stream, Visited, Succ);
+  }
+}
+
+void graph_impl::printGraphAsDot(const char *GraphName) {
+  std::string GraphNameStr(GraphName);
+  std::string FileName = "graph_" + GraphNameStr + ".dot";
+
+  std::fstream Stream(FileName, std::ios::out);
+  Stream << "strict digraph {" << std::endl;
+
+  MVisitedNodes.clear();
```

Comment:
Can we just define `std::vector<std::shared_ptr<node_impl>> VisitedNodes;` here inside this function, rather than making `MVisitedNodes` a graph member variable?